actors/miner/src/deadline_state.rs

```diff
@@ -102,6 +103,92 @@ impl Deadlines {
         self.due[deadline_idx as usize] = store.put_cbor(deadline, Code::Blake2b256)?;
         Ok(())
     }
+
+    pub fn move_partitions<BS: Blockstore>(
+        store: &BS,
+        orig_deadline: &mut Deadline,
+        dest_deadline: &mut Deadline,
+        partitions: &BitField,
+    ) -> anyhow::Result<()> {
+        let mut orig_partitions = orig_deadline.partitions_amt(store)?;
+        let mut dest_partitions = dest_deadline.partitions_amt(store)?;
+
+        // even though we're moving partitions intact, we still need to update orig/dest `Deadline` accordingly.
+
+        let first_dest_partition_idx = dest_partitions.count();
+        for (i, orig_partition_idx) in partitions.iter().enumerate() {
+            let moving_partition = orig_partitions
+                .get(orig_partition_idx)?
+                .ok_or_else(|| actor_error!(not_found, "no partition {}", orig_partition_idx))?
+                .clone();
+            if !moving_partition.faults.is_empty() || !moving_partition.unproven.is_empty() {
+                return Err(actor_error!(forbidden, "partition with faults or unproven sectors are not allowed to move, partition_idx {}", orig_partition_idx))?;
+            }
+            if !moving_partition.faulty_power.is_zero() {
+                return Err(actor_error!(
+                    illegal_state,
+                    "partition faulty_power should be zero when faults is empty, partition_idx {}",
+                    orig_partition_idx
+                ))?;
+            }
+
+            let dest_partition_idx = first_dest_partition_idx + i as u64;
+
+            let all_sectors = moving_partition.sectors.len();
+            let live_sectors = moving_partition.live_sectors().len();
+            if orig_deadline.early_terminations.get(orig_partition_idx) {
+                return Err(actor_error!(forbidden, "partition with early terminated sectors are not allowed to move, partition_idx {}", orig_partition_idx))?;
+            }
+
+            // start updating orig/dest `Deadline` here
+
+            orig_deadline.total_sectors -= all_sectors;
+            orig_deadline.live_sectors -= live_sectors;
+
+            dest_deadline.total_sectors += all_sectors;
+            dest_deadline.live_sectors += live_sectors;
+
+            orig_partitions.set(orig_partition_idx, Partition::new(store)?)?;
+            dest_partitions.set(dest_partition_idx, moving_partition)?;
+        }
+
+        // update expirations_epochs Cid of Deadline.
+        {
+            let mut epochs_to_remove = Vec::<u64>::new();
+            let mut orig_expirations_epochs: Array<BitField, _> =
+                Array::load(&orig_deadline.expirations_epochs, store)?;
```

Comment:
It seems `BitfieldQueue` doesn't help here, and here we're modifying two `Array<BitField, _>` within `for_each_mut`, which is also referring to contextual variables like `partitions`,`first_dest_partition_idx` etc, moving to a method will make it more cluttering..   I put a comment there to avoid the confusion.